formats/avro/src/main/java/io/cloudevents/avro/AvroFormat.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package io.cloudevents.avro;
+
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.CloudEventData;
+import io.cloudevents.SpecVersion;
+import io.cloudevents.core.builder.CloudEventBuilder;
+import io.cloudevents.core.data.BytesCloudEventData;
+import io.cloudevents.core.format.EventDeserializationException;
+import io.cloudevents.core.format.EventFormat;
+import io.cloudevents.core.format.EventSerializationException;
+import io.cloudevents.core.v1.CloudEventV1;
+import io.cloudevents.rw.CloudEventDataMapper;
+import io.cloudevents.types.Time;
+import io.cloudevents.v1.avro.CloudEvent.Builder;
+
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * An implementation of {@link EventFormat} for <a href="github.com/cloudevents/spec/blob/v1.0.1/avro-format">the Avro format</a>.
+ * This format is resolvable with {@link io.cloudevents.core.provider.EventFormatProvider} using the content type {@link #AVRO_CONTENT_TYPE}.
+ * It only supports data that is bytes.
+ */
+public class AvroFormat implements EventFormat {
+
+  public static final String AVRO_CONTENT_TYPE = "application/cloudevents+avro";
+
+  @Override
+  public byte[] serialize(CloudEvent ce) throws EventSerializationException {
+    try {
+      Builder builder;
+      builder = io.cloudevents.v1.avro.CloudEvent.newBuilder();
+
+      Map<String, Object> attribute = new HashMap<String,Object>() {
+        @Override
+        public Object put(String key, Object value) {
+          if (value instanceof Boolean || value instanceof Integer || value instanceof String)
```

Comment:
But those 'are' CloudEvent attribute types .. so an application can (and arguable should) create a CloudEvent using code like this ...

```
OffsetDateTime someTime = ...... ;
URI someURI = .....

builder.withExtension('myattrname', someTime);
builder.withExtension('anotherattr', someURI);
```

When that gets presented to your serializer it will be an _OffsetDateTime_ (or _URI_) not a _String_

I'm fairly confident in my assertion, the proof would be to use one of the pre-canned CloudEvent objects for the test-data code as those use events are constructed using the CE Types.

Reply:
BTW .. It's absolutely true that JSON doesn't have these types, but the spec does, and so does the 'information model' for CE so those other types can be presented at the time of serialization (the proto and xml format specifications allow for the type information to be preserved during transit).

Reply:
Avro does not have the types either.

Reply:
I agree with the sentiment and that type information does get dropped today in some situations, but we should be able to marshal those value types into their String representation if they get presented IMHO ..

Reply:
I’ve refactored the code, so this is not supported unintentionally.